File: go/parser.go
package amino

import (
	"regexp"
	"strconv"
	"strings"
)

// ── tokenizer ──────────────────────────────────────────────────────────────

var fixedSymbols = []string{">=", "<=", "!=", ">", "<", "=", "(", ")", "[", "]", ",", "."}

var (
	tokFloatRe = regexp.MustCompile(`^-?\d+\.\d+`)
	tokIntRe   = regexp.MustCompile(`^-?\d+`)
	tokStrRe   = regexp.MustCompile(`^'[^']*'`)
	tokIdentRe = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*`)
)

func tokenize(text string, opSymbols []string) ([]string, error) {
	symbols := append([]string{}, opSymbols...)
	symbols = append(symbols, fixedSymbols...)
	// Longest match first.
	for i := 1; i < len(symbols); i++ {
		for j := i; j > 0 && len(symbols[j]) > len(symbols[j-1]); j-- {
			symbols[j], symbols[j-1] = symbols[j-1], symbols[j]
		}
	}
	var tokens []string
	i := 0
	for i < len(text) {
		ch := text[i]
		if ch == ' ' || ch == '\t' {
			i++
			continue
		}
		rest := text[i:]
		if strings.HasPrefix(rest, "not in") {
			tokens = append(tokens, "not in")
			i += 6
			continue
		}
		if m := tokStrRe.FindString(rest); m != "" {
			tokens = append(tokens, m)
			i += len(m)
			continue
		}
		if m := tokFloatRe.FindString(rest); m != "" {
			tokens = append(tokens, m)
			i += len(m)
			continue
		}
		if m := tokIntRe.FindString(rest); m != "" {
			tokens = append(tokens, m)
			i += len(m)
			continue
		}
		if m := tokIdentRe.FindString(rest); m != "" {
			tokens = append(tokens, m)
			i += len(m)
			continue
		}
		matched := false
		for _, sym := range symbols {
			if strings.HasPrefix(rest, sym) {
				tokens = append(tokens, sym)
				i += len(sym)
				matched = true
				break
			}
		}
		if !matched {
			return nil, newError(CodeSyntax, "unexpected character %q at position %d", string(ch), i)
		}
	}
	return tokens, nil
}

// ── parser ─────────────────────────────────────────────────────────────────

type parser struct {
	tokens   []string
	pos      int
	schema   *Schema
	ops      *operatorRegistry
	types    *typeRegistry
	depth    int
	maxDepth int
	builtin  map[string]bool // every built-in operator token, enabled or not
}

func (p *parser) peek() (string, bool) {
	if p.pos < len(p.tokens) {
		return p.tokens[p.pos], true
	}
	return "", false
}

func (p *parser) advance() (string, error) {
	if p.pos >= len(p.tokens) {
		return "", newError(CodeSyntax, "unexpected end of expression")
	}
	t := p.tokens[p.pos]
	p.pos++
	return t, nil
}

func (p *parser) leftBP(tok string) int {
	bp, ok := p.ops.bindingPower(tok)
	if !ok {
		return 0
	}
	return bp
}

func (p *parser) parse() (Node, error) {
	node, err := p.parseExpr(0)
	if err != nil {
		return nil, err
	}
	if tok, ok := p.peek(); ok {
		if p.builtin[tok] {
			return nil, newError(CodeUnknownOperator, "operator %q is not enabled", tok)
		}
		return nil, newError(CodeSyntax, "unexpected token %q", tok)
	}
	return node, nil
}

func (p *parser) enter() error {
	p.depth++
	if p.depth > p.maxDepth {
		return newError(CodeDepthExceeded, "expression nesting exceeds %d", p.maxDepth)
	}
	return nil
}

func (p *parser) parseExpr(minBP int) (Node, error) {
	if err := p.enter(); err != nil {
		return nil, err
	}
	defer func() { p.depth-- }()

	left, err := p.nud()
	if err != nil {
		return nil, err
	}
	for {
		tok, ok := p.peek()
		if !ok || tok == ")" || tok == "]" || tok == "," {
			break
		}
		twoTokenNotIn := tok == "not" && p.pos+1 < len(p.tokens) && p.tokens[p.pos+1] == "in"
		if twoTokenNotIn {
			tok = "not in"
		}
		bp := p.leftBP(tok)
		if bp <= minBP {
			break
		}
		p.pos++
		if twoTokenNotIn {
			p.pos++
		}
		left, err = p.led(tok, left)
		if err != nil {
			return nil, err
		}
	}
	return left, nil
}

func (p *parser) nud() (Node, error) {
	tok, err := p.advance()
	if err != nil {
		return nil, err
	}

	if tok == "(" {
		node, err := p.parseExpr(0)
		if err != nil {
			return nil, err
		}
		if t, ok := p.peek(); !ok || t != ")" {
			return nil, newError(CodeSyntax, "expected ')'")
		}
		p.pos++
		return node, nil
	}

	if tok == "[" {
		items, err := p.parseListBody()
		if err != nil {
			return nil, err
		}
		return &Literal{Value: items, Type: "List"}, nil
	}

	if def := p.ops.prefix(tok); def != nil {
		operand, err := p.parseExpr(def.BindingPower)
		if err != nil {
			return nil, err
		}
		if !p.compatible(def.InputTypes[0], operand.TypeName()) {
			return nil, newError(CodeTypeMismatch, "operator %q expects %s, got %s", tok, def.InputTypes[0], operand.TypeName())
		}
		return &Unary{Op: tok, Operand: operand, Type: def.ReturnType, Def: def}, nil
	}

	if tokFloatRe.MatchString(tok) && len(tokFloatRe.FindString(tok)) == len(tok) {
		f, _ := strconv.ParseFloat(tok, 64)
		return &Literal{Value: f, Type: "Float"}, nil
	}
	if tokIntRe.MatchString(tok) && len(tokIntRe.FindString(tok)) == len(tok) {
		i, err := strconv.ParseInt(tok, 10, 64)
		if err != nil {
			return nil, newError(CodeSyntax, "integer literal out of range: %s", tok)
		}
		return &Literal{Value: i, Type: "Int"}, nil
	}
	if len(tok) >= 2 && tok[0] == '\'' && tok[len(tok)-1] == '\'' {
		return &Literal{Value: tok[1 : len(tok)-1], Type: "Str"}, nil
	}
	if tok == "true" {
		return &Literal{Value: true, Type: "Bool"}, nil
	}
	if tok == "false" {
		return &Literal{Value: false, Type: "Bool"}, nil
	}

	if tokIdentRe.MatchString(tok) && len(tokIdentRe.FindString(tok)) == len(tok) {
		if t, ok := p.peek(); ok && t == "(" {
			return p.parseCall(tok)
		}
		name := tok
		for {
			t, ok := p.peek()
			if !ok || t != "." {
				break
			}
			p.pos++
			part, err := p.advance()
			if err != nil {
				return nil, err
			}
			name += "." + part
		}
		f := p.schema.Lookup(name)
		if f == nil {
			return nil, newError(CodeUnknownField, "unknown field %q", name)
		}
		return &Variable{Name: name, Type: f.TypeName}, nil
	}

	if p.builtin[tok] {
		return nil, newError(CodeSyntax, "operator %q has no left operand", tok)
	}
	return nil, newError(CodeSyntax, "unexpected token %q", tok)
}

func (p *parser) led(tok string, left Node) (Node, error) {
	right, err := p.parseExpr(p.leftBP(tok))
	if err != nil {
		return nil, err
	}
	def, err := p.resolve(tok, left, right)
	if err != nil {
		return nil, err
	}
	return &Binary{Op: tok, Left: left, Right: right, Type: def.ReturnType, Def: def}, nil
}

// resolve picks the operator definition for the operand types, or reports a
// type mismatch. Exact type-name matches win; then definitions whose declared
// input types are compatible by base type; wildcard definitions apply the
// built-in compatibility rules.
func (p *parser) resolve(tok string, left, right Node) (*OperatorDef, error) {
	lt, rt := left.TypeName(), right.TypeName()
	cands := p.ops.candidates(tok)
	if len(cands) == 0 {
		return nil, newError(CodeUnknownOperator, "no operator %q", tok)
	}
	for _, d := range cands {
		if len(d.InputTypes) == 2 && d.InputTypes[0] == lt && d.InputTypes[1] == rt {
			return d, nil
		}
	}
	for _, d := range cands {
		if len(d.InputTypes) != 2 {
			continue
		}
		if p.compatible(d.InputTypes[0], lt) && p.compatible(d.InputTypes[1], rt) {
			if d.InputTypes[0] == "*" && d.InputTypes[1] == "*" {
				if err := p.checkBuiltin(tok, lt, rt, right); err != nil {
					return nil, err
				}
			}
			if d.InputTypes[1] == "List" {
				if err := p.checkMembership(tok, lt, right); err != nil {
					return nil, err
				}
			}
			return d, nil
		}
	}
	return nil, newError(CodeTypeMismatch, "no operator %q for types (%s, %s)", tok, lt, rt)
}

// compatible reports whether a value of type actual can be passed where the
// operator declares want.
func (p *parser) compatible(want, actual string) bool {
	if want == "*" {
		return true
	}
	wb, ab := p.baseOf(want), p.baseOf(actual)
	if wb == ab {
		return true
	}
	return isNumericBase(wb) && isNumericBase(ab)
}

func (p *parser) baseOf(name string) string {
	if strings.HasPrefix(name, "List[") || name == "List" {
		return "List"
	}
	return p.types.base(name)
}

func isNumericBase(b string) bool { return b == "Int" || b == "Float" }

// checkBuiltin applies the rules for the wildcard-typed comparison operators.
func (p *parser) checkBuiltin(tok, lt, rt string, right Node) error {
	lb, rb := p.baseOf(lt), p.baseOf(rt)
	numeric := isNumericBase(lb) && isNumericBase(rb)
	switch tok {
	case "=", "!=":
		if lb == rb || numeric {
			return nil
		}
	case ">", "<", ">=", "<=":
		if numeric || (lb == "Str" && rb == "Str") {
			return nil
		}
	default:
		return nil
	}
	return newError(CodeTypeMismatch, "cannot apply %q to %s and %s", tok, lt, rt)
}

// checkMembership requires a list literal on the right whose elements match
// the left operand's type.
func (p *parser) checkMembership(tok, lt string, right Node) error {
	lit, ok := right.(*Literal)
	if !ok || lit.Type != "List" {
		return newError(CodeTypeMismatch, "%q needs a list literal on the right", tok)
	}
	items, _ := lit.Value.([]any)
	if len(items) == 0 {
		return nil
	}
	eb := literalBase(items[0])
	lb := p.baseOf(lt)
	if lb == eb || (isNumericBase(lb) && isNumericBase(eb)) {
		return nil
	}
	return newError(CodeTypeMismatch, "cannot test %s membership in a list of %s", lt, eb)
}

func literalBase(v any) string {
	switch v.(type) {
	case bool:
		return "Bool"
	case int64:
		return "Int"
	case float64:
		return "Float"
	case string:
		return "Str"
	case []any:
		return "List"
	}
	return "?"
}

func (p *parser) parseCall(name string) (Node, error) {
	p.pos++ // (
	var args []Node
	for {
		t, ok := p.peek()
		if !ok {
			return nil, newError(CodeSyntax, "unexpected end of expression in call to %q", name)
		}
		if t == ")" {
			break
		}
		arg, err := p.parseExpr(0)
		if err != nil {
			return nil, err
		}
		args = append(args, arg)
		if t, ok := p.peek(); ok && t == "," {
			p.pos++
		}
	}
	p.pos++ // )
	fn := p.schema.LookupFunction(name)
	if fn == nil {
		return nil, newError(CodeUnknownFunction, "unknown function %q", name)
	}
	required := 0
	for _, prm := range fn.Params {
		if !prm.Optional {
			required++
		}
	}
	if len(args) < required || len(args) > len(fn.Params) {
		return nil, newError(CodeTypeMismatch, "function %q takes %d to %d arguments, got %d", name, required, len(fn.Params), len(args))
	}
	for i, arg := range args {
		if !p.compatible(fn.Params[i].TypeName, arg.TypeName()) {
			return nil, newError(CodeTypeMismatch, "function %q argument %d expects %s, got %s", name, i+1, fn.Params[i].TypeName, arg.TypeName())
		}
	}
	return &Call{Name: name, Args: args, Type: fn.ReturnType}, nil
}

func (p *parser) parseListBody() ([]any, error) {
	items := []any{}
	for {
		t, ok := p.peek()
		if !ok {
			return nil, newError(CodeSyntax, "unterminated list literal")
		}
		if t == "]" {
			break
		}
		v, err := p.parseLiteralValue()
		if err != nil {
			return nil, err
		}
		items = append(items, v)
		if t, ok := p.peek(); ok && t == "," {
			p.pos++
		}
	}
	p.pos++ // ]
	return items, nil
}

func (p *parser) parseLiteralValue() (any, error) {
	tok, err := p.advance()
	if err != nil {
		return nil, err
	}
	if tok == "[" {
		if err := p.enter(); err != nil {
			return nil, err
		}
		defer func() { p.depth-- }()
		return p.parseListBody()
	}
	if len(tok) >= 2 && tok[0] == '\'' && tok[len(tok)-1] == '\'' {
		return tok[1 : len(tok)-1], nil
	}
	if tok == "true" {
		return true, nil
	}
	if tok == "false" {
		return false, nil
	}
	if tokFloatRe.MatchString(tok) && len(tokFloatRe.FindString(tok)) == len(tok) {
		f, _ := strconv.ParseFloat(tok, 64)
		return f, nil
	}
	if tokIntRe.MatchString(tok) && len(tokIntRe.FindString(tok)) == len(tok) {
		i, err := strconv.ParseInt(tok, 10, 64)
		if err != nil {
			return nil, newError(CodeSyntax, "integer literal out of range: %s", tok)
		}
		return i, nil
	}
	return nil, newError(CodeSyntax, "expected literal, got %q", tok)
}

func parseExpression(text string, e *Engine) (Node, error) {
	if len(text) > e.maxLength {
		return nil, newError(CodeSyntax, "expression longer than %d characters", e.maxLength)
	}
	tokens, err := tokenize(strings.TrimSpace(text), e.ops.symbolList())
	if err != nil {
		return nil, err
	}
	p := &parser{tokens: tokens, schema: e.schema, ops: e.ops, types: e.types, maxDepth: e.maxDepth, builtin: builtinTokens()}
	return p.parse()
}

func builtinTokens() map[string]bool {
	out := map[string]bool{}
	for _, d := range standardOperators() {
		out[d.Token()] = true
	}
	return out
}

File: typescript/src/parser.ts
import type { Binary, Call, Literal, LiteralValue, Node, Unary, Variable } from "./ast.js";
import { AminoError } from "./errors.js";
import type { OperatorRegistry, ResolvedOperator } from "./operators.js";
import { builtinTokens } from "./operators.js";
import type { Schema } from "./schema.js";
import type { TypeRegistry } from "./types.js";

const FIXED_SYMBOLS = [">=", "<=", "!=", ">", "<", "=", "(", ")", "[", "]", ",", "."];
const FLOAT = /^-?\d+\.\d+/;
const INT = /^-?\d+/;
const STR = /^'[^']*'/;
const IDENT = /^[a-zA-Z_][a-zA-Z0-9_]*/;
const FULL_FLOAT = /^-?\d+\.\d+$/;
const FULL_INT = /^-?\d+$/;
const FULL_IDENT = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

export function tokenize(text: string, opSymbols: string[]): string[] {
  const symbols = [...opSymbols, ...FIXED_SYMBOLS].sort((a, b) => b.length - a.length);
  const tokens: string[] = [];
  let i = 0;
  while (i < text.length) {
    const ch = text[i] as string;
    if (ch === " " || ch === "\t") {
      i++;
      continue;
    }
    const rest = text.slice(i);
    if (rest.startsWith("not in")) {
      tokens.push("not in");
      i += 6;
      continue;
    }
    const m = STR.exec(rest) ?? FLOAT.exec(rest) ?? INT.exec(rest) ?? IDENT.exec(rest);
    if (m) {
      tokens.push(m[0]);
      i += m[0].length;
      continue;
    }
    const sym = symbols.find((s) => rest.startsWith(s));
    if (sym === undefined) throw new AminoError("syntax", `unexpected character "${ch}" at position ${i}`);
    tokens.push(sym);
    i += sym.length;
  }
  return tokens;
}

export interface ParserContext {
  schema: Schema;
  ops: OperatorRegistry;
  types: TypeRegistry;
  maxDepth: number;
  maxLength: number;
}

class Parser {
  private pos = 0;
  private depth = 0;
  private readonly builtin = builtinTokens();

  constructor(
    private readonly tokens: string[],
    private readonly ctx: ParserContext,
  ) {}

  private peek(): string | undefined {
    return this.tokens[this.pos];
  }

  private advance(): string {
    const t = this.tokens[this.pos];
    if (t === undefined) throw new AminoError("syntax", "unexpected end of expression");
    this.pos++;
    return t;
  }

  private leftBP(tok: string): number {
    return this.ctx.ops.bindingPower(tok) ?? 0;
  }

  parse(): Node {
    const node = this.parseExpr(0);
    const tok = this.peek();
    if (tok !== undefined) {
      if (this.builtin.has(tok)) throw new AminoError("unknown_operator", `operator "${tok}" is not enabled`);
      throw new AminoError("syntax", `unexpected token "${tok}"`);
    }
    return node;
  }

  private enter(): void {
    this.depth++;
    if (this.depth > this.ctx.maxDepth) {
      throw new AminoError("depth_exceeded", `expression nesting exceeds ${this.ctx.maxDepth}`);
    }
  }

  private parseExpr(minBP: number): Node {
    this.enter();
    try {
      let left = this.nud();
      for (;;) {
        let tok = this.peek();
        if (tok === undefined || tok === ")" || tok === "]" || tok === ",") break;
        const twoTokenNotIn = tok === "not" && this.tokens[this.pos + 1] === "in";
        if (twoTokenNotIn) tok = "not in";
        const bp = this.leftBP(tok);
        if (bp <= minBP) break;
        this.pos += twoTokenNotIn ? 2 : 1;
        left = this.led(tok, left);
      }
      return left;
    } finally {
      this.depth--;
    }
  }

  private nud(): Node {
    const tok = this.advance();

    if (tok === "(") {
      const node = this.parseExpr(0);
      if (this.peek() !== ")") throw new AminoError("syntax", "expected ')'");
      this.pos++;
      return node;
    }

    if (tok === "[") {
      const value = this.parseListBody();
      return { node: "literal", value, type: "List" } satisfies Literal;
    }

    const prefix = this.ctx.ops.prefix(tok);
    if (prefix) {
      const operand = this.parseExpr(prefix.bindingPower);
      const want = prefix.inputTypes[0] ?? "*";
      if (!this.compatible(want, operand.type)) {
        throw new AminoError("type_mismatch", `operator "${tok}" expects ${want}, got ${operand.type}`);
      }
      return { node: "unary", op: tok, operand, type: prefix.returnType, def: prefix } satisfies Unary;
    }

    if (FULL_FLOAT.test(tok)) return { node: "literal", value: parseFloat(tok), type: "Float" };
    if (FULL_INT.test(tok)) return { node: "literal", value: parseInt(tok, 10), type: "Int" };
    if (tok.length >= 2 && tok.startsWith("'") && tok.endsWith("'")) {
      return { node: "literal", value: tok.slice(1, -1), type: "Str" };
    }
    if (tok === "true") return { node: "literal", value: true, type: "Bool" };
    if (tok === "false") return { node: "literal", value: false, type: "Bool" };

    if (FULL_IDENT.test(tok)) {
      if (this.peek() === "(") return this.parseCall(tok);
      let name = tok;
      while (this.peek() === ".") {
        this.pos++;
        name += "." + this.advance();
      }
      const field = this.ctx.schema.lookup(name);
      if (!field) throw new AminoError("unknown_field", `unknown field "${name}"`);
      return { node: "variable", name, type: field.typeName } satisfies Variable;
    }

    if (this.builtin.has(tok)) throw new AminoError("syntax", `operator "${tok}" has no left operand`);
    throw new AminoError("syntax", `unexpected token "${tok}"`);
  }

  private led(tok: string, left: Node): Node {
    const right = this.parseExpr(this.leftBP(tok));
    const def = this.resolve(tok, left, right);
    return { node: "binary", op: tok, left, right, type: def.returnType, def } satisfies Binary;
  }

  private resolve(tok: string, left: Node, right: Node): ResolvedOperator {
    const lt = left.type;
    const rt = right.type;
    const cands = this.ctx.ops.candidates(tok);
    if (cands.length === 0) throw new AminoError("unknown_operator", `no operator "${tok}"`);
    for (const d of cands) {
      if (d.inputTypes.length === 2 && d.inputTypes[0] === lt && d.inputTypes[1] === rt) return d;
    }
    for (const d of cands) {
      if (d.inputTypes.length !== 2) continue;
      const [w0, w1] = d.inputTypes as [string, string];
      if (this.compatible(w0, lt) && this.compatible(w1, rt)) {
        if (w0 === "*" && w1 === "*") this.checkBuiltin(tok, lt, rt);
        if (w1 === "List") this.checkMembership(tok, lt, right);
        return d;
      }
    }
    throw new AminoError("type_mismatch", `no operator "${tok}" for types (${lt}, ${rt})`);
  }

  private compatible(want: string, actual: string): boolean {
    if (want === "*") return true;
    const wb = this.baseOf(want);
    const ab = this.baseOf(actual);
    return wb === ab || (isNumericBase(wb) && isNumericBase(ab));
  }

  private baseOf(name: string): string {
    if (name === "List" || name.startsWith("List[")) return "List";
    return this.ctx.types.base(name);
  }

  private checkBuiltin(tok: string, lt: string, rt: string): void {
    const lb = this.baseOf(lt);
    const rb = this.baseOf(rt);
    const numeric = isNumericBase(lb) && isNumericBase(rb);
    switch (tok) {
      case "=":
      case "!=":
        if (lb === rb || numeric) return;
        break;
      case ">":
      case "<":
      case ">=":
      case "<=":
        if (numeric || (lb === "Str" && rb === "Str")) return;
        break;
      default:
        return;
    }
    throw new AminoError("type_mismatch", `cannot apply "${tok}" to ${lt} and ${rt}`);
  }

  private checkMembership(tok: string, lt: string, right: Node): void {
    if (right.node !== "literal" || right.type !== "List") {
      throw new AminoError("type_mismatch", `"${tok}" needs a list literal on the right`);
    }
    const items = right.value as LiteralValue[];
    const first = items[0];
    if (first === undefined) return;
    const eb = literalBase(first);
    const lb = this.baseOf(lt);
    if (lb === eb || (isNumericBase(lb) && isNumericBase(eb))) return;
    throw new AminoError("type_mismatch", `cannot test ${lt} membership in a list of ${eb}`);
  }

  private parseCall(name: string): Call {
    this.pos++;
    const args: Node[] = [];
    for (;;) {
      const t = this.peek();
      if (t === undefined) throw new AminoError("syntax", `unexpected end of expression in call to "${name}"`);
      if (t === ")") break;
      args.push(this.parseExpr(0));
      if (this.peek() === ",") this.pos++;
    }
    this.pos++;
    const fn = this.ctx.schema.lookupFunction(name);
    if (!fn) throw new AminoError("unknown_function", `unknown function "${name}"`);
    const required = fn.params.filter((p) => !p.optional).length;
    if (args.length < required || args.length > fn.params.length) {
      throw new AminoError("type_mismatch", `function "${name}" takes ${required} to ${fn.params.length} arguments, got ${args.length}`);
    }
    args.forEach((arg, i) => {
      const param = fn.params[i] as (typeof fn.params)[number];
      if (!this.compatible(param.typeName, arg.type)) {
        throw new AminoError("type_mismatch", `function "${name}" argument ${i + 1} expects ${param.typeName}, got ${arg.type}`);
      }
    });
    return { node: "call", name, args, type: fn.returnType };
  }

  private parseListBody(): LiteralValue[] {
    const items: LiteralValue[] = [];
    for (;;) {
      const t = this.peek();
      if (t === undefined) throw new AminoError("syntax", "unterminated list literal");
      if (t === "]") break;
      items.push(this.parseLiteralValue());
      if (this.peek() === ",") this.pos++;
    }
    this.pos++;
    return items;
  }

  private parseLiteralValue(): LiteralValue {
    const tok = this.advance();
    if (tok === "[") {
      this.enter();
      try {
        return this.parseListBody();
      } finally {
        this.depth--;
      }
    }
    if (tok.length >= 2 && tok.startsWith("'") && tok.endsWith("'")) return tok.slice(1, -1);
    if (tok === "true") return true;
    if (tok === "false") return false;
    if (FULL_FLOAT.test(tok)) return parseFloat(tok);
    if (FULL_INT.test(tok)) return parseInt(tok, 10);
    throw new AminoError("syntax", `expected literal, got "${tok}"`);
  }
}

function isNumericBase(b: string): boolean {
  return b === "Int" || b === "Float";
}

function literalBase(v: LiteralValue): string {
  if (typeof v === "boolean") return "Bool";
  if (typeof v === "number") return Number.isInteger(v) ? "Int" : "Float";
  if (typeof v === "string") return "Str";
  return "List";
}

export function parseExpression(text: string, ctx: ParserContext): Node {
  if (text.length > ctx.maxLength) {
    throw new AminoError("syntax", `expression longer than ${ctx.maxLength} characters`);
  }
  const tokens = tokenize(text.trim(), ctx.ops.symbols());
  return new Parser(tokens, ctx).parse();
}
